### PR TITLE
Code Review, Pull Requests team standards documentation

### PR DESCRIPTION
**Pull request by rcantin-w**
## What does this change?
Reorganises the "Github process" section and add recommendations for the Pull Request/Code review processes.

This is very much up for review and changes based on the team's opinions, consider it a first draft.

If you have questions/ideas about new _processes_, add them to this document: https://docs.google.com/document/d/1vh95JrnJWOdFn4AE9kPDBWUGIWv9DQh4WCgxurZ8BXU/edit

### CODE REVIEW

File: developer-handbook/github-process/pull-request-process.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Pull Request creation
+
+Here are recommendations to consider from the moment you create a new branch to work on.
+
+## Pre-creation considerations
+
+**Consider creating your Pull Request when you're 30-50% through**
+_If it's a big piece_, consider creating your PR when you're 30-50% through. Create it as a draft, so no one gets notified of every commit.
```
Comment (kenoir) [inline developer-handbook/github-process/pull-request-process.md]: What about suggesting creation in draft "as soon as possible", with a focus on filling out the PR description up front? 

PRs flag that work is being done and provide context for other developers. Short, contextless descriptions created after most of the work are done can feel like they make things quicker, but potentially disadvantage anyone trying to review, or understand the work later.

File: developer-handbook/github-process/code-review-process.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	- Duplication of code (e.g. helpers that we already have a library for)
+	- Edge cases
+	- Leftovers from development, (e.g. `console.log`, addressed `TODO`s)
+	- Typos!
+
+- If you see something cool, or a nice detail, point it out. Positive comments on a PR are always appreciated.
+
+- If the review becomes more of a conversation, ensure you get back to it promptly when new commits are submitted. 
+
+- Use "Request changes" sparingly. We don't really want to use it, but if you think the PR contains security gaps, for example, it might be best to flag it as a change request.
```
Comment (kenoir) [inline developer-handbook/github-process/code-review-process.md:35]: Agree. My view is that "request changes" is for when a PR is going to break something (or a security issue like you point out), anything else can be worked out by "comment only", with the expectation that comments are going to be addressed before merge.

File: developer-handbook/github-process/post-code-review.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# Post Code Review
+
+Your Pull Request was reviewed by a colleague (or a few colleagues, which is great). What's left to do now is
+
+## Address it
+- The "keep your comments friendly" rule applies here too! We all have the best intentions for our codebases.
+- React or reply to every comment that was left, even if it's just "acknowledged" or "done", or an emoji reaction. The person that left it thought it was worth saying something, it's nice to acknowledge it.
```
Comment (kenoir) [inline developer-handbook/github-process/post-code-review.md:8]: 👍 Comments can be marked "resolved" to indicate that they've been addressed whether action is taken or not & responding "not addressing this here, but in a future PR" can be a valid reply if it makes sense in context.

File: developer-handbook/github-process/code-review-process.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+- Ensure you follow the "How to test" section, and run your own tests on top of it if you can think of other useful ones.
+
+- Keep your comments friendly; we all know we are reviewing the code and not the coder, but the lines are easily blurred (for example, try not to use words like "simply", "easily", "just", "obviously", "well, actually...").
+
+- Code review is as much a priority as writing code is, so ensure you code review promptly to unblock your colleagues: make it part of your routine. For example, you could have a look at [the PR listing page](https://github.com/pulls/review-requested) in the morning, or in between pieces of work, to see if anything needs reviewing. If you can't find time, try to set a time aside in the future and let your colleagues know when that will be.
+
+- If it's a huge PR, consider doing it in bits instead of in one go, it will be more efficient for your focus. Take your time!
+
+- Don't be afraid to ask questions! It's a prime place to learn, and it might mean that the code could be more readable, or use comments.
```
Comment (kenoir) [inline developer-handbook/github-process/code-review-process.md:11]: 👌

File: developer-handbook/github-process/pull-request-process.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+If they aren't as urgent or related to the original ticket, they don't need to be addressed in your PR (and probably shouldn't be). But every `TODO` should have a Github issue linked, so create one and add its link as a comment if it's not the case. Ensure it's on the team's radar if it's a bigger piece of work. 
+
+## Pull Request creation considerations
+A pull request should have a clear description with:
+  - Linked ticket, which is where the context should live
+  - Changes made
+  - How to test
+  - Risks and plan to mitigate
+
+If useful, leave comments throughout the code before making it "Ready for review", to add to the context of the changes. If they are important, or useful for future devs, it should be left as comments in the code instead.
```
Comment (kenoir) [inline developer-handbook/github-process/pull-request-process.md:37]: Yeah! I sometimes like reviewing my own PR and adding comments with context or changes i'd like to make.

File: developer-handbook/github-process/post-code-review.md (hunk 3 above)
Comment (rcantin-w) [reply developer-handbook/github-process/post-code-review.md:8]: At my other jobs, the person that left the original comment would be the one with the responsibility to Resolve it once they considered it addressed (either in that PR or by creating a ticket). Therefore, they would be part of their review and mark the steps towards Approval. I didn't know if that might be considered too much?

File: developer-handbook/github-process/code-review-process.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+	- Readability
+	- Compliance (e.g. a11y)
+	- Dead code
+	- Performance concerns
+	- Security concerns
+	- Gaps in logic (e.g. in an `if` statement, is every case covered?)
+	- Duplication of code (e.g. helpers that we already have a library for)
+	- Edge cases
+	- Leftovers from development, (e.g. `console.log`, addressed `TODO`s)
+	- Typos!
```
Comment (rcantin-w) [inline developer-handbook/github-process/code-review-process.md:26]: Do the tests cover all the cases?

File: developer-handbook/github-process/code-review-process.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+- If you see something cool, or a nice detail, point it out. Positive comments on a PR are always appreciated.
```
Comment (paul-butcher) [inline developer-handbook/github-process/code-review-process.md:31]: This is a nice detail.

File: developer-handbook/github-process/code-review-process.md (hunk 6 above)
Comment (paul-butcher) [reply developer-handbook/github-process/code-review-process.md:26]: Some thoughts on writing/reviewing tests (probably too much for this section, but could be boiled down a bit)
* Do they adequately reflect the questions asked in the ticket?
* Do they cover appropriate boundary conditions? (a common shortcoming is that tests tend to cover what you want to happen, but skip over what happens when the input is "wrong")
* Do the tests tell a story? Are they easily readable?
* Do they actually test what they say they are testing?

File: developer-handbook/github-process/code-review-process.md (hunk 2 above)
Comment (paul-butcher) [reply developer-handbook/github-process/code-review-process.md:35]: It's nice to see this point stated explicitly.  It reflects how I feel about it and use "Request changes", and given how infrequently I've seen it here (if at all), it appears to be how we use it as a team.  But I don't think anyone has ever said this out loud in my presence, so I have sometimes wondered whether my approach is correct.

File: developer-handbook/github-process/code-review-process.md [hunk 8]
```diff
@@ -0,0 +1,36 @@
+# Code review
+
+You've been assign a Pull Request to review, and you want to get started.
```
Comment (paul-butcher) [inline developer-handbook/github-process/code-review-process.md]:
```suggestion
You've been assigned a Pull Request to review, and you want to get started.
```

File: developer-handbook/github-process/code-review-process.md [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+- If it's a huge PR, consider doing it in bits instead of in one go, it will be more efficient for your focus. Take your time!
+
+- Don't be afraid to ask questions! It's a prime place to learn, and it might mean that the code could be more readable, or use comments.
+
+- If a comment is not a blocker for release, make sure you state that by prefixing it with "Nit Pick".
+
+- Don't be a perfectionist, but keep an eye out for:
+	- Our own chosen standards (naming conventions, file structure, etc.)
+	- Readability
```
Comment (paul-butcher) [inline developer-handbook/github-process/code-review-process.md:17]: This one (along with the tests) is my own top priority when reviewing. I think this is the part that is most likely to be missed by a developer who is really looking closely at a problem.

File: developer-handbook/github-process/post-code-review.md (hunk 3 above)
Comment (paul-butcher) [reply developer-handbook/github-process/post-code-review.md:8]: When It's my work being reviewed, I like to use "resolved" as a sort of checklist of things I have to do.  Particularly if there are a few identical points in different places - (e.g. "typo" ... "and here"... "same again here") where they do not (all) warrant a reply.

If the baton gets handed back to the reviewer after changes, that can be an opportunity for the reviewer to say "This one hasn't actually been resolved"

File: developer-handbook/github-process/post-code-review.md [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+- If some comments/suggestions need more discussion, switch to Slack or a call, but ensure you document the outcome in the PR for legacy's sake.
```
Comment (paul-butcher) [inline developer-handbook/github-process/post-code-review.md]: And link between the PR and the conversation if possible.

File: developer-handbook/github-process/pull-request-process.md [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+This also ensures that the lottery factor isn't a problem; if you can't be the one to finish the work, others have eyes on it.
+
+**Pull Requests should be small as possible**
+If you're working on something big, consider splitting it in multiple PRs. If that's not possible, aim to split your commits logically, with meaningful commit messages.
+
+**Address `TODOs`**
+If there are `TODO`s in the changed files, even if they were there before you started your work: it's your responsability to check they have a chance to get addressed. 
+
+If they aren't as urgent or related to the original ticket, they don't need to be addressed in your PR (and probably shouldn't be). But every `TODO` should have a Github issue linked, so create one and add its link as a comment if it's not the case. Ensure it's on the team's radar if it's a bigger piece of work.
```
Comment (paul-butcher) [inline developer-handbook/github-process/pull-request-process.md]: Consider whether the TODO belongs in a PR comment, rather than a code comment.

i.e. are you writing the TODO so that a reviewer understands that you know something should be different but is not necessarily pertinent to the task at hand?

Comment (rcantin-w) [reply developer-handbook/github-process/pull-request-process.md]: To double-check this point, if a TODO belongs to the PR, do we mean that the PR shouldn't get merged until it gets addressed?

File: developer-handbook/github-process/code-review-process.md (hunk 7 above)
Comment (rcantin-w) [reply developer-handbook/github-process/code-review-process.md:31]: I see what you've done there.

File: developer-handbook/github-process/post-code-review.md (hunk 3 above)
Comment (rcantin-w) [reply developer-handbook/github-process/post-code-review.md:8]: I agree, when it's something small like a typo, it doesn't need a second opinion! And like you said, otherwise it's nice to have confirmation you got the recommended change right.

File: developer-handbook/github-process/pull-request-process.md (hunk 11 above)
Comment (paul-butcher) [reply developer-handbook/github-process/pull-request-process.md]: I think "addressed" in this context means "fixed or a separate ticket has been written".  Whether leaving it for a separate ticket/PR is acceptable can be up to the reviewers to agree/disagree.

The sentiment expressed by raising a (non-draft) PR is that the author thinks it's good to go, so any comments by the author should mean "This might not be right, but I think it's good enough".  However, as part of self-review, one might put comments in a draft PR that do need to be fixed before merging, and might be overlooked when pressing the "Ready for Review" button.

The sentiment expressed in a TODO is "I know this is imperfect/incomplete, but I haven't done anything about it yet".  I think it's better to state this in a review comment and linking it to a separate ticket, with the meaning "I'm not going to fix it in this PR, here's the ticket to fix it later", rather than in a code comment.

The problem with TODOs and ticket links in code comments is that they get stale in weird ways, whereas if they are in PR comments, we can always see the actual context.

File: developer-handbook/github-process/pull-request-process.md (hunk 1 above)
Comment (agnesgaroux) [reply developer-handbook/github-process/pull-request-process.md]: How about "Consider creating your Pull Request as soon as possible. Think of PRs less as validation of work done and more as visibility of work in progress. The description ideally offers context as to why the work is being done as well as explaining how the solution was decided on, which can be very helpful to reviewers. Create it as a draft, so no one gets notified of every commit."

File: developer-handbook/github-process/post-code-review.md (hunk 3 above)
Comment (agnesgaroux) [reply developer-handbook/github-process/post-code-review.md:8]: Yes, unless it's a trivial change the original reviewer/commenter should have a chance to check out how the feedback was addressed

File: developer-handbook/github-process/pull-request-process.md (hunk 1 above)
Comment (rcantin-w) [reply developer-handbook/github-process/pull-request-process.md]: Love that, thank you!

File: developer-handbook/github-process/post-code-review.md (hunk 3 above)
Comment (rcantin-w) [reply developer-handbook/github-process/post-code-review.md:8]: I'm adding this below 

> The owner of the PR can "resolve" comments after commenting "done" if they are small things like typos. For bigger changes than that, the responsibility of resolving should be passed on the the reviewer as they can use it to confirm the changes are inline with what they'd suggested.

Does that suit?

File: developer-handbook/github-process/code-review-process.md (hunk 6 above)
Comment (rcantin-w) [reply developer-handbook/github-process/code-review-process.md:26]: I'm adding this in the list ([change can be seen here](https://github.com/wellcomecollection/developer-handbook/pull/1/files#diff-cb8ddfad2df336eced53ef91175fe84c923d686b44f468699b28d7b900d7d37fR27))
> - Do the tests cover all cases? Do they actually test what they say they are testing?

But I liked the other questions for the "Pull request process" as they're a good indicator of what tests should be like! [So added it as a list here](https://github.com/wellcomecollection/developer-handbook/pull/1/files#diff-98802f63a1e1ae0bb73422d2c806ebd85275584be6ecf804d2b03e002481b74dR34-R37)

> - Would new tests be relevant? If so, add them:
>    - Do they adequately reflect the questions asked in the ticket?
>    - Do they cover appropriate boundary conditions? (tests tend to cover what you want to happen, but should also cover what happens when the input is "wrong")
>    - Do the tests tell a story? Are they easily readable?

what do you think?

File: developer-handbook/github-process/pull-request-process.md (hunk 11 above)
Comment (rcantin-w) [reply developer-handbook/github-process/pull-request-process.md]: I understand what you mean, we definitely have 3year old+ TODOs in the code. One of the "pros" was that they were findeable that way, whereas a comment in a PR that is 3 years old shall probably never be seen again. The important thing is then just to ensure a ticket is created and in a backlog somewhere.

In-code TODOs also add context to a piece of code I might see 6 months later that I find a bit wonky, and give me permission to refactor?

But I think you're saying both scenarios _are_ valid, I'm just unsure how to phrase it. Could you help me?

File: developer-handbook/github-process/pull-request-process.md (hunk 1 above)
Comment (rcantin-w) [reply developer-handbook/github-process/pull-request-process.md]: I've made changes to the whole section, [see here](https://github.com/wellcomecollection/developer-handbook/pull/1/files#diff-98802f63a1e1ae0bb73422d2c806ebd85275584be6ecf804d2b03e002481b74dR7-R12)

File: developer-handbook/github-process/post-code-review.md (hunk 3 above)
Comment (rcantin-w) [reply developer-handbook/github-process/post-code-review.md:8]: Added [here](https://github.com/wellcomecollection/developer-handbook/pull/1/files#diff-1e074398a9aa14bdce2e9c63512c5f37cbb8a8463c79f72c58e10483258412fbR10)

**Review by kenoir — APPROVED** [review]

**Review by georgiaewhitney — APPROVED** [review]
Kept reading through changes and comments, but didn't review - sorry! 
Looks and reads great - thanks for this work.

**Review by gestchild — APPROVED** [review]
Looks great. Thank you for doing this

**Review by StepanBrychta — APPROVED** [review]
Looks great!

**Review by agnesgaroux — APPROVED** [review]
I approve 👍

File: developer-handbook/github-process/github-process.md [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+* PRs reference a set of git commits which in totality describe a coherent functional change.
+* Issues reference one or more PRs which together describe a coherent functional change.
+* Github Projects contain Issues _or_ PRs which transition through defined states towards _done_.
+
+## Processes 
+As a team, we operate in a very friendly, trusted, no-blame culture. The nature of code review can sometimes make this feel a little bit difficult, especially if you're a new joiner.
+
+That's why we hope documenting our standards for both pull request and code review processes will help in strengthening our confidence and interpretation of them. We also hope it helps empower new joiners in tackling their first code reviews, or submitting their first feature.
+
+We're approaching this from a place of believing code review to be an excellent a learning tool. Both for the submitter, who benefits from suggestions on their PR, but also for the reviewer. It's a prime place to ask questions, and learn from the submitter themselves.
```
Comment (gestchild) [inline developer-handbook/github-process/github-process.md]: Just noticed an extra 'a' between 'excellent' and 'learning'

File: developer-handbook/github-process/post-code-review.md [hunk 13]
```diff
@@ -0,0 +1,19 @@
+# Post Code Review
+
+Your Pull Request was reviewed by a colleague (or a few colleagues, which is great). What's left to do now is:
+
+## Address it
+- Just like in the code review step, we opt for friendly tones in their replies as well! We all have the best intentions for our codebases and every input is valuable.
```
Comment (gestchild) [inline developer-handbook/github-process/post-code-review.md:6]: should this be 'in _the_ replies as well'?

Comment (rcantin-w) [reply developer-handbook/github-process/post-code-review.md:6]: I meant it like "the code reviews' replies", but am open to correcting as I don't know which one is correct? 😅

Comment (gestchild) [reply developer-handbook/github-process/post-code-review.md:6]: Ah, that makes sense too. I just didn't read it that way. Happy to keep it as is 😄

File: developer-handbook/github-process/post-code-review.md [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+
+- React or reply to every comment that was left, even if it's just "acknowledged" or "done", or an emoji reaction. The person that left it thought it was worth saying something, it's nice to acknowledge it.
+
+- The owner of the PR can "resolve" comments after commenting "done" if they are small things like typos. For bigger changes than that, the responsibility of resolving should be passed on the the reviewer as they can use it to confirm the changes are inline with what they'd suggested.
```
Comment (gestchild) [inline developer-handbook/github-process/post-code-review.md]: 'passed on the the reviewer' should be 'passed on to the reviewer' I think

File: developer-handbook/github-process/github-process.md (hunk 12 above)
Comment (rcantin-w) [reply developer-handbook/github-process/github-process.md]: addressed, resolving convo

File: developer-handbook/github-process/post-code-review.md (hunk 14 above)
Comment (rcantin-w) [reply developer-handbook/github-process/post-code-review.md]: Addressed, resolving.

File: developer-handbook/github-process/post-code-review.md (hunk 13 above)
Comment (rcantin-w) [reply developer-handbook/github-process/post-code-review.md:6]: OK! Resolving

File: developer-handbook/github-process/pull-request-process.md [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+This will also let others gain some familiarity with the code changes, and might allow them to flag things in advance. 
+
+Another important aspect of this is that it ensures that the lottery factor isn't a problem; if you can't be the one to finish the work, others have eyes on it, and access to the latest code changes.
+
+**Pull Requests should be small as possible**
+
+If you're working on something big, consider splitting it in multiple PRs. If that's not possible, aim to split your commits logically, with meaningful commit messages.
+
+**Address `TODOs`**
```
Comment (davidpmccormick) [inline developer-handbook/github-process/pull-request-process.md]: Nit `TODO`s!

**Review by davidpmccormick — APPROVED** [review]
IMG_7229

File: developer-handbook/github-process/pull-request-process.md (hunk 15 above)
Comment (rcantin-w) [reply developer-handbook/github-process/pull-request-process.md]: fixed! Resolving

File: developer-handbook/github-process/pull-request-process.md (hunk 11 above)
Comment (kenoir) [reply developer-handbook/github-process/pull-request-process.md]: > The important thing is then just to ensure a ticket is created and in a backlog somewhere.

Agree with this, over adding TODO comments to code. My feeling is that we should avoid adding any "TODO" at all to `main`, instead:

- Create a ticket to do the work that can be prioritised
- Fix the issue before merging
- Add a comment describing the peculiarity or optimisation for context for the next person to read without implying the next person should fix it. In this context I am not against adding links to issues in GitHub!

In my opinion the issue is a TODO adds a "here is a bit of work you should do later" without any sense of priority or severity of the problem it indicates.  Adding an issue to the backlog allows the work to be properly prioritised in view of other devs/delivery/product people.

Comment (rcantin-w) [reply developer-handbook/github-process/pull-request-process.md:23]: I do like that.... If the others are good to go for that I'd put that in.

Comment (rcantin-w) [reply developer-handbook/github-process/pull-request-process.md]: (amendment to prior comment😄 ) I do like that, very action focussed. If no one is objecting to this, I'll put that in.

Comment (rcantin-w) [reply developer-handbook/github-process/pull-request-process.md]: Now amended, section changed completely.
